README.md
# Naive Bayes for Sentiment Analysis

### Group members:
- Konstantin
- Yegor 
---

## Project Background

The [Bayes' theorem](https://en.wikipedia.org/wiki/Bayes%27_theorem) is a simple yet powerful mathematical formula that allows predictions to be made based on prior beliefs and the likelihood of evidence observed.

<img src="/resources/bayes_formula_dark.svg#gh-light-mode-only" width="220px" alt="Bayes' theorem'">
<img src="/resources/bayes_formula_light.svg#gh-dark-mode-only" width="220px" alt="Bayes' theorem'">
 
- $A, B = \text{events} $
- $P(A) = \text{independent probability of A}$
- $P(B) = \text{independent probability of B}$
- $P(A|B) = \text{probability of A given B is true}$
- $P(B|A) = \text{probability of B given A is true}$

A [Naive Bayes classifier](https://en.wikipedia.org/wiki/Naive_Bayes_classifier) is in the family of probabilistic classifiers that is used among other things in tasks such as: [spam prediction](https://en.wikipedia.org/wiki/Naive_Bayes_spam_filtering#:~:text=Naive%20Bayes%20spam%20filtering%20is,with%20roots%20in%20the%201990s), sentiment analysis, recommendation systems.

## What problem does our program solve?
Here we will be using the algorithm for sentiment analysis for movie reviews to classify them as either a positive or a negative review.

Using [Haskell](https://wiki.haskell.org/Haskell), our program will:
- perform **text preprocessing** on text data.
  - remove stop words (words that don't add much value to the text).
  - remove symbols/punctuation.
- train and test the model using a publicly available dataset[^1] - the [IMDb dataset](https://ai.stanford.edu/~amaas/data/sentiment/) contains 25,000 positive and 25,000 negative movie reviews collected from [IMDb](https://en.wikipedia.org/wiki/IMDb).
  - although the user can use our application on any other data, as long as it is added as a txt file in the correct folder within the *datasets* directory.
- perform validation of a trained model using test data.
- implement a command-line interface where users can input strings of reviews and get the model's inference result.

## What is the something extra?

In addition to the expected basic functionality, our program will:
- allow users to download the model parameters in a text file.
- re-train the model with various dataset sizes and compare their performance.

## What did we learn from doing this?

We acquired more practical experience in writing Haskell applications and Cabal projects. We learned a more in-depth about Input/Output, several data structures and their operations (Map,Set, etc). We also acquired a better intuition about cryptic Haskell compile errors and how to debug them. Also, not the least important, we got a taste of working with files, like reading the data to and from them. Finally, by implementing the simple Naive-Bayes algorithm we saw the potential in implementing machine learning algorithms in Haskell.

---

## How to use this program?

Assuming all imports are working with cabal (which is unlikely since we haven't tested it and we are sure there are problems), just do:
1. Go into /app folder
2. Type `ghci`
3. Type `:l Main`
4. Type `main` to start the Interactive CLI

How to use the CLI?
- Once your program is running with `main`, the `help` command gives all of the commands that you need to know
- You can exit by typing `exit` or `e`
- You can get the list of all models in the /models folder by typing `models` or `m`
- You can select/load a model from that list using `load <model_name`
- You can see the currently selected model and some info about it using `state` or `s`
- You can train a new model using `train <model_name>`. If a duplicate name is given, it overwrites that model I believe
- You can validate a model using `validate` or `v`, it always validates the currently selected model (we didn't have time to implement more functions so bear with us)

## What does the program do?
When doing training, the program is able to generate new training and testing sets, since it shuffles the file lists randomly.
You can select the training split, we made 3 options:
- 25%-75% train-validation %
- 50%-50% train-validation %
- 75%-25% train-validation %

---
- You don't need to validate right away with our program, since we save the test data sets for later use :)
- The program does not re-validate a model that was previously validated, as long as the its score can be found in scores.json 
  - This saves a lot of time for unnecessary recomputation

[^1]: dataset citation, see [also](https://github.com/yegory/NaiveBayes/tree/master/datasets): 
@InProceedings{maas-EtAl:2011:ACL-HLT2011,
  author    = {Maas, Andrew L.  and  Daly, Raymond E.  and  Pham, Peter T.  and  Huang, Dan  and  Ng, Andrew Y.  and  Potts, Christopher},
  title     = {Learning Word Vectors for Sentiment Analysis},
  booktitle = {Proceedings of the 49th Annual Meeting of the Association for Computational Linguistics: Human Language Technologies},
  month     = {June},
  year      = {2011},
  address   = {Portland, Oregon, USA},
  publisher = {Association for Computational Linguistics},
  pages     = {142--150},
  url       = {http://www.aclweb.org/anthology/P11-1015 }
}


CHANGELOG.md
# Revision history for NaiveBayes

## 0.1.0.0 -- 2023-02-22
Set-up

## 0.1.0.1 -- 2023-02-23
Added ability to read the text (String) in txt file inside a directory, and do this for every single txt file.
The result can be stored in a <String> array 

## 1.1.0.0 -- 2023-02-26
Added some CLI functionality, a lot of functions are moved around. 
Most of the previous usage for the program has been changed hence the version update.

TODO for next version:
1) Perform validation on dataset
2) Add ability to train a model using random review data (25-75, 50-50, 25-75 %train-test splits)
    - e.g. come up with a way to use random number generator to create an array of randomly picked files to use for training.
    - the test files will just be the left over files.
    - maybe need to encorporate it into some data structure to capture the state of the CLI program, because
    - the idea is that the user can train and test at a later time after several load (or save, train..) operations 
    - (but perhaps this is too much and maybe just better to make validation run always after training?)
3) Add some way for the user to input their own text into the console and have the currently selected trained model give a verdict if their sentence is positive or negative (with  some certainty score or verification score) 

app/Main.hs
{- ======================================= IMPORTS ======================================= -}
import Interactive
import Tools
import System.IO

-- https://hackage.haskell.org/package/filepath-1.4.100.1/docs/System-FilePath-Posix.html
import System.FilePath 
import qualified Data.Text as T
import qualified Data.Text.IO as TIO
import Data.Map (Map)
import Data.Set (Set)
import Data.Maybe (Maybe)
import qualified Data.Map as Map
import qualified Data.Set as Set
import qualified Data.Maybe as Maybe

punctuation = "!\"#$%&'()*+, -./:;<=>?@[]^_`{|}~ "

-- backlog of possible TODOS: ( who has time for this LOL? )
-- PERFORM VALIDATION
-- FILE IS NOT EMPTY, what should happen?
-- FILE EXISTS, read the test,

{- ======================================== MAIN ======================================== -}


main :: IO ()
main 
    = do 
        start
        --- TRAIN THE MODEL
        -- accuracy <- trainAndValidate

        -- comment below out if you want
        -- let fileName = "example"
        -- writeMapModelToDiskAsJSON fileName map_train_pos
        -- s <- readJsonModelFromDisk fileName
        -- printMap (Map.assocs s)
        -- start

        putStr "" -- prevent errors in case no IO is performed 


-- create the vocabulary set by combining keys from positive and negative maps
-- perform laplace smoothing 
-- convert to log probabilities
-- output the model as a tuple of the pos and neg log prob dicts and the vocabulary


{- ======================================= HELPERS ======================================= -}


-- Helper function to call printFileContents on each path in filePaths 
printDirContents :: [String] -> IO ()
printDirContents [] = putStr ""
printDirContents filePaths@(x:xs)
    = do
        printFileContents x
        printDirContents xs
-- Try in main:
-- txtFilePaths <- getAllTxtFilePaths trainPositivePath :: IO [String] -- create arraypath of each file
-- printDirContents txtFilePaths -- print contents of each file



-- Possible errors:
-- the openFile computation may fail with isAlreadyInUseError if the file is already open and cannot be reopened; 
-- isDoesNotExistError if the file does not exist; or isPermissionError if the user does not have permission to open the file.
printFileContents :: FilePath -> IO ()
printFileContents filePath 
    = do
        -- openFile allocates and returns a new open handle to manage the file in filePath
        -- Opening in ReadMode, the handle is positioned at the beginning
        hdl <- openFile filePath ReadMode
        contents <- hGetContents hdl
        putStrLn contents -- returns the list of characters corresponding to the unread portion of the channel or file managed by hdl
        
        hClose hdl -- Closing the file to avoid causing isAlreadyInUseError in the future


printMap :: [(String, Int)] -> IO ()
printMap [] = putStr ""
printMap kvpair@((a, b):xs)
    = do
        putStrLn (a ++ " : " ++ show b)
        printMap xs

app/Interactive.hs
module Interactive where

import RandomShuffle
import System.IO
import Text.Read   (readMaybe)
import Tools
import Data.Set (Set)
import Data.Map (Map)
import qualified Data.Map as Map
import qualified Data.Set as Set
import qualified Data.Maybe as Maybe
-- To run it, try:
-- ghci Interactive
-- start

{- ====================================== DATA DEF ====================================== -}

welcomeText = "\n~\tHi human!\n~\tWelcome to the Naive Bayes application\n\n"
helpPrompt = "Type a command; 'help' for a list of commands; or 'exit' to exit: "
exitMessage = "\n\t--------------------------- Bye! Bye! ---------------------------\n"
listOfCommands = ["List of commands:",
                  "‣ Enter 'state' to get info about the current model selected (if any)",
                  "‣ Enter 'train <val> <model_name>' to train a model with the given name\n  | <val> can be one of:\n  |  A := 25%-75% train-test split\n  |  B := 50%-50% train-test split\n  |  C := 75%-25% train-test split",
                  "‣ Enter 'models' for a list of existing <model_name>`s",
                  "‣ Enter 'load <model_name>' to load an existing model",
                  "‣ Enter 'validate' to validate the currently selected model"]

-- need
-- array of saved models
-- 

{- ======================================== MAIN ======================================== -}

data State = Empty
            |  Model (Maybe LoadedModel) ValidationScore
        deriving (Ord, Eq, Show)

type LoadedModel = (ModelName, ModelMap)
type ModelMap = (Map String Double, Map String Double, Set String)
type ModelName = String
type ValidationScore = Maybe Double



start :: IO ()
start
    = do
        putStr welcomeText
        let emptyState = Model Nothing Nothing
        main_loop emptyState


main_loop :: State -> IO ()
main_loop state =
   do
        line <- getResponse helpPrompt
        jsonFileNames <- getAllModelNames
        let lline = lowerString $ trim line
        case lline of
            v   |   s `elem` ["exit", "e"]                                                           -> exit
                |   s `elem` ["state", "s"]                                                          -> printStateMaybe state
                |   s `elem` ["help", "h"] && n == 2 && (head t) `elem` ["train", "t"]               -> trainHelp state
                |   s `elem` ["help", "h"]                                                           -> printHelp state
                |   s `elem` ["models", "m"]                                                         -> printModels state
                |   s `elem` ["train", "t"] && n == 3 && lowerString (head t) `elem` ["a", "b", "c"] -> trainModelHandler state (lowerString (head t)) (head (tail t))
                |   s `elem` ["train", "t"] && lowerString (head t) `elem` ["help", "h"]             -> trainHelp state
                |   s `elem` ["train", "t"]                                                          -> trainError state     
                |   s `elem` ["load", "l"] && n == 2 && head t `elem` jsonFileNames                  -> loadModelHandler state (head t) 
                |   s `elem` ["load", "l"] && n == 2                                                 -> loadErrorModelNotFound state (head t) 
                |   s `elem` ["load", "l"] && n > 2                                                  -> loadErrorTooManyArguments state (n - 1)
                |   s `elem` ["load", "l"]                                                           -> loadErrorNotEnoughArguments state
                |   s `elem` ["validate", "v"]                                                       -> validateCurrentModelHandler state
                |   otherwise -> do
                                putStrLn ("\n~  Sorry, but the command `" ++ lline ++ "` was not found! \n~  Type help for list of commands")
                                main_loop state
                where   (s, t) = (head split, tail split) :: (String, [String])
                        split = (splitOn " " v) :: [String]
                        n = length split -- number of total arguments including the command s

{- ======================================= HELPERS ======================================= -}

exit :: IO ()
exit
    = do
        putStr exitMessage


getResponse :: String -> IO String
getResponse question
    = do
        putStr question
        getLine

printHelp :: State -> IO ()
printHelp state
    = do
        putStr "\n"
        printStringArray listOfCommands
        putStr "\n"
        main_loop state



printStateMaybe :: State -> IO ()
printStateMaybe state@(Model maybeModel validationScore)
    = do
        case maybeModel of
            Just (modelName, modelMap) -> printState modelName modelMap validationScore
            Nothing -> putStrLn "\n~  Currently no model is selected :(\n~  Hang on though! You can train a new model or select an existing one!\n"
        main_loop state


printState :: String -> (Map String Double, Map String Double, Set String) -> ValidationScore -> IO ()
printState modelName modelMap@(m1, m2, set) validationScore
    = do
        putStrLn ("Model's file name (without extension): " ++ modelName)
        putStrLn ("Trained positive map contains: " ++ show (Map.size m1) ++ " unique words (keys)")
        putStrLn ("Trained negative map contains: " ++ show (Map.size m2) ++ " unique words (keys)")
        putStrLn ("Number of unique words in both maps: " ++ show (length set) ++ " unique words (keys)")

        -- TODO: add total count of words maybe or some interesting statistic?
        case validationScore of
            Just validationScore -> putStrLn ("And the validation score is: " ++ (show validationScore))
            Nothing -> putStrLn "The model currently does not have a validation score, try the 'validate' command!"



printModels :: State -> IO ()
printModels state
    = do
        jsonFileNames <- getAllModelNames
        printStringArray jsonFileNames
        main_loop state



-- A := 25%-75% train-test split\n B := 50%-50% train-test split\n  |  C := 75%-25% train-test split
trainChoiceToRatio :: String -> IO Double
trainChoiceToRatio s
    = do
        case s of 
            v   | v == "a" -> return 0.25
                | v == "b" -> return 0.50
                | v == "c" -> return 0.75
                | otherwise -> do
                                putStrLn "Error in trainChoiceToRatio (choice not of a, b, or c)"
                                return 0.5


trainModelHandler :: State -> String -> String -> IO ()
trainModelHandler state@(Model maybeModel validationScore) option modelName
    = do
        ratio <- trainChoiceToRatio option
        trainP_FilePaths <- getAllTxtFilePaths trainPositivePath :: IO [String]
        trainN_FilePaths <-getAllTxtFilePaths trainNegativePath :: IO [String]
        testP_FilePaths <- getAllTxtFilePaths testPositivePath :: IO [String] 
        testN_FilePaths <-getAllTxtFilePaths testNegativePath :: IO [String]

        let positivePaths = trainP_FilePaths ++ testP_FilePaths
        let negativePaths = trainN_FilePaths ++ testN_FilePaths

        positiveShuffled <- shuffle positivePaths -- all positive paths that are shuffled
        negativeShuffled <- shuffle negativePaths
        
        let len_positivePaths = length positivePaths
        let len_negativepPaths = length negativePaths

        let numOfPosFilesToTake = floor (fromIntegral len_positivePaths * ratio)
        let numOfNegFilesToTake = floor (fromIntegral len_negativepPaths * ratio)
    
        let pos_ShuffledTrain = take numOfPosFilesToTake positiveShuffled
        let neg_ShuffledTrain = take numOfNegFilesToTake negativeShuffled
        let pos_ShuffledTest = drop numOfPosFilesToTake positiveShuffled
        let neg_ShuffledTest = drop numOfNegFilesToTake negativeShuffled

        s <- readFileStrict stopwordsFilePath :: IO String
        let stopWords = splitOn ['\n'] s :: [String]

        map_train_pos <- getDirContents Map.empty stopWords pos_ShuffledTrain :: IO (Map String Int)
        map_train_neg <- getDirContents Map.empty stopWords neg_ShuffledTrain :: IO (Map String Int)

        let map_tp = Map.delete "" map_train_pos 
        let map_tn = Map.delete "" map_train_neg

        model <- trainModel map_tp map_tn :: IO (Map String Double, Map String Double, Set String)

        writeTrainedModelToDiskAsJSON modelName model
        writeValidationFilePathArraysToDiskAsJSON modelName (pos_ShuffledTest, neg_ShuffledTest)
        -- putStrLn (show totalLength)
        -- printStringArray shuffledPathIndices
        
        case maybeModel of
            Just (modelName, modelMap) -> main_loop state
            Nothing -> main_loop (Model (Just (modelName, model)) Nothing)
      


-- triggered when a valid model file with a base name of modelName was provided by user
loadModelHandler :: State -> String -> IO ()
loadModelHandler state modelName 
    = do
        -- we need to load the current model into state
        -- putStrLn modelName
        readModel <- readTrainedModelFromDisk modelName :: IO (Maybe (Map String Double, Map String Double, Set String))
        case readModel of
            Nothing -> do
                        putStrLn "Warning: the loaded model is empty!"
                        main_loop state
            Just model -> do   
                                putStrLn "Model read!"
                                saved_score <- getModelEvaluationIfExists modelName :: IO (Maybe Double)     
                                case saved_score of
                                    Nothing -> main_loop (Model (Just (modelName, model)) Nothing)
                                    Just score -> main_loop (Model (Just (modelName, model)) (Just score))



validateCurrentModelHandler :: State -> IO () 
validateCurrentModelHandler state@(Model maybeModel validationScore)
    = do
        case validationScore of
            -- if validationScore exists already, it means that this model was evaluated so just output it
            Just score -> putStrLn ("The model evaluated test files with " ++ show (score) ++ " accuracy!")
            Nothing -> 
                do
                    case maybeModel of
                        -- if the model exists at least, then
                        Just a@(modelName, model) -> do -- validate current model
                                                saved_score <- getModelEvaluationIfExists modelName :: IO (Maybe Double)
                                                case saved_score of
                                                    Just value -> putStrLn ("The model evaluated test files with " ++ show (value) ++ " accuracy!")
                                                    Nothing -> do
                                                            validationPaths <- readJsonValidationFilePathArrays modelName :: IO (Maybe ([FilePath], [FilePath]))
                                                            case validationPaths of
                                                                Just (pos_paths, neg_paths) -> do
                                                                    stopWords <- getStopWords
                                                                    score <- validationAccuracy model pos_paths neg_paths stopWords :: IO Double
                                                                    writeModelEvaluation modelName score
                                                                    putStrLn ("The model evaluated on the test files with " ++ show (score) ++ " accuracy!")
                                                                    let new_state = Model (Just a) (Just score)
                                                                    main_loop new_state
                                                                Nothing -> do
                                                                        putStrLn "Model training failed previously as couldn't find validation paths"
                                                                        main_loop state
                        -- if the model doesn't exist
                        Nothing -> do
                                putStrLn "\tLoad or train a model before performing validation!\n"
                                main_loop state

trainHelp :: State -> IO ()
trainHelp state
    = do
        putStrLn "Enter 'train <val>' to start training a model\n\t<val> can be one of A, B, or C:\n\t| A := 25%-75% train-test split\n\t| B := 50%-50% train-test split\n\t| C := 75%-25% train-test split\n"
        main_loop state

loadHelp :: State -> IO ()
loadHelp state
    = do
        putStrLn "Enter 'load <model_name>' to load an existing model\nEnter 'models' for a list of existing <model_name>`s\n"
        main_loop state

validateHelp :: State -> IO ()
validateHelp state
    = do
        putStrLn "Enter 'validate' to validate the currently selected model, or 'validate <model_name>' to validate a specific model.\n     Gives an error if no model is selected or <model_name> is not in \"./models\" foldern\n"
        main_loop state

trainError :: State -> IO ()
trainError state 
    = do
        putStrLn "\n~  BadNumberOfArgumentsError: Expected 1 argument after 'train' representing the train mode. Type 'help train' for information about the 'train' subcommand\n"           
        main_loop state

loadErrorNotEnoughArguments :: State -> IO ()
loadErrorNotEnoughArguments state 
    = do
        putStrLn "\n~  BadNumberOfArgumentsError: Expected 1 argument after 'load', but got 0\n"     
        main_loop state

loadErrorTooManyArguments :: State -> Int -> IO ()
loadErrorTooManyArguments state nArgs 
    = do
        putStrLn ("\n~  TooManyArgumentsError: Expected 1 argument after 'load', but got " ++ show nArgs ++ "\n")
        main_loop state

loadErrorModelNotFound :: State -> String -> IO ()
loadErrorModelNotFound state model
    = do
        putStrLn ("\n~  ModelNotFoundError: The model '" ++ model ++ "' was not found in the in the \"./models\" folder.\n")
        main_loop state